ic-factory/src/state.rs

```diff
@@ -1,139 +1,455 @@
     }
 
-    pub fn set_icp_to(&mut self, to: Principal) -> Result<(), FactoryError> {
-        self.check_controller_access()?;
-        self.configuration.icp_to = to;
+    /// Number of canisters the factory keeps track of.
+    pub fn canister_count(&self) -> usize {
+        self.canisters.len()
+    }
 
-        Ok(())
+    /// List of canisters the factory keeps track of.
+    pub fn canister_list(&self) -> Vec<Principal> {
+        self.canisters.keys().copied().collect()
+    }
+
+    /// HashMap of canisters the factory keeps track of with their code hashes.
+    pub fn canisters(&self) -> &HashMap<Principal, CanisterHash> {
+        &self.canisters
     }
 
+    /// Locks the `FactoryState`, prohibiting any update operations on it until the returned lock
+    /// object is dropped. See [`UpdateLock`] documentation for more details about how and why this works.
+    pub fn lock(&mut self) -> Result<UpdateLock, FactoryError> {
+        self.update_lock.lock()
+    }
+
+    fn check_update_allowed(&self) -> Result<(), FactoryError> {
+        match self.update_lock.is_locked() {
+            true => Err(FactoryError::StateLocked),
+            false => Ok(()),
+        }
+    }
+
+    // If the caller can provide a lock to this method, it means that the state is locked, as there
+    // is no way to get the lock object except by calling the [`lock`] method. So the purpose of
+    // this check is to guard against creating a completely different `UpdateLock` object and
+    // giving it to a facotry method. In such case we simply panick to make it clear that the code
+    // that did such a thing is broken and must be fixed.
+    fn check_lock(&self, lock: &UpdateLock) {
+        assert_eq!(*lock, self.update_lock, "invalid update lock usage")
+    }
+
+    /// Consumes the fee for canister creation in the form of cycles (if provided by the call) or
+    /// ICP in other case. Returns an error in case nor cycles nor ICP are provided and the caller
+    /// is not the factory controller.
     pub fn consume_provided_cycles_or_icp(
         &self,
         caller: Principal,
-        // ) -> Pin<Box<dyn Future<Output = Result<u64, FactoryError>> + Send>> {
-    ) -> Pin<Box<dyn Future<Output = Result<u64, FactoryError>>>> {
-        Box::pin(consume_provided_cycles_or_icp(
-            caller,
-            self.ledger_principal(),
-            self.icp_to(),
-            self.icp_fee(),
-            self.controller(),
+    ) -> impl Future<Output = Result<u64, FactoryError>> {
+        let ledger = self.ledger_principal();
+        let icp_to = self.icp_to();
+        let icp_fee = self.icp_fee();
+        let controller = self.controller();
+
+        consume_provided_cycles_or_icp(caller, ledger, icp_to, icp_fee, controller)
+    }
+}
+
+/// Abstraction to provided compile time checks for factory method access.
+pub struct Authorized<'a, T> {
+    factory: &'a mut FactoryState,
+    phantom: PhantomData<T>,
```

Comment:
I think the intention was to have a typesafe flag for `Authorized`, e.g. allowing extending it with `User` or similar structs. This will mean that `User` will have the same `factory` field as `Owner`, but a separate API. This'll bothersome since with your proposal `User` and `Owner` will have the similar `factory` fields which should be equal in any case.

Reply:
Agreed. Then remove it if it's not used anywhere

Reply:
But maybe it's a good compile time check for restricting calling owner-related functions such as `set_canister_wasm` so that we won't be needed prefixing it with `self.validate_owner()?` or something like that. I think we should leave it as is.